### PR TITLE
Implement the `signature-bridge` contract

### PR DESCRIPTION
**Pull request by duguorong009**
Summary of changes
Changes introduced in this pull request

Implement the `signature-bridge` contract
   - Add 2 execution entries(`admin_set_res_with_signature` & `exec_proposal_with_signature`)
   - Add 1 query(`get_config`)

Reference issue to close (if applicable)
Closes

### CODE REVIEW

File: contracts/signature-bridge/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        proposal_nonce: state.proposal_nonce,
+    })
+}
+
+// Verifying signature of governor over some datahash
+fn signed_by_governor(data: &[u8], sig: &[u8], governor: &str) -> Result<bool, ContractError> {
+    let hashed_data = Keccak256::hash(data).map_err(|_| ContractError::HashError)?;
+    let message = Message::parse(&hashed_data);
+    let sig = Signature::parse_standard_slice(sig).map_err(|_| ContractError::HashError)?;
+    let recv_id = RecoveryId::parse(0_u8).map_err(|_| ContractError::HashError)?;
```
Comment (drewstone) [inline contracts/signature-bridge/src/contract.rs:200]: This looks incorrect. Please refer to https://docs.rs/sp-io/4.0.0/src/sp_io/lib.rs.html#770-793.

The `recv_id` should come from the signature data itself.

**Review by drewstone — CHANGES_REQUESTED** [review]
Good start! Can we rename things like `res` and `RES` and `Res` to `resource`, `RESOURCE`, and `Resource`? It should be clear we're setting resources and not anything else. Also looks like the ecdsa recover method will be incorrect since you default to 0 recv id. I would check how to parse that value from the signature itself (it exists on there as the last byte afaik but double check the documentation of the other repos or the libsecp256k1 repo.

File: contracts/signature-bridge/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    if msg.nonce != state.proposal_nonce + 1 {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "Invalid nonce".to_string(),
+        }));
+    }
+
+    let func_sig = Keccak256::hash(
+        b"adminSetResourceWithSignature(bytes32,bytes4,uint32,bytes32,address,address,bytes)",
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply contracts/signature-bridge/src/contract.rs:99]: There is no concept equivalent to `function signature` here in `cosmwasm`.
Simply, `cosmwasm` is based on `wasm(web assembly)`, which is a black box.
It only provides some pre-defined entries like `instantiate`, `execute`, `query` etc, in order to communicate with external world.
For more info, you can check the followings:
https://github.com/CosmWasm/cosmwasm/blob/main/README.md
https://github.com/CosmWasm/cosmwasm/blob/main/EntryPoints.md

Hence, I think the `function signature` logic is unnecessary, here in `cosmwasm` contracts.

File: contracts/signature-bridge/src/contract.rs (hunk 1 above)
Comment (duguorong009) [reply contracts/signature-bridge/src/contract.rs]: Replaced the `libsecp256` lib with the cosmwasm-builtin `secp256k1_verify` util.
https://github.com/webb-tools/protocol-cosmwasm/pull/37/commits/2b845e7ee8119bba81d3ff3b693de047f55ba077

File: contracts/signature-bridge/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+    // Parse resourceID from the data
+    let resource_id_bytes = &msg.data[0..32];
+    let resource_id = element_encoder(resource_id_bytes);
+
+    // Parse chain ID + chain type from the resource ID
+    let execution_chain_id_type: u64 = get_chain_id_type(&resource_id_bytes[26..32]);
+
+    // Verify current chain matches chain ID from resource ID
+    if compute_chain_id_type(MOCK_CHAIN_ID, &COSMOS_CHAIN_TYPE) != execution_chain_id_type {
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply contracts/signature-bridge/src/contract.rs:166]: Add the `NOTE` comments for this code block.
Looks like needs more discussion for this part.

File: contracts/signature-bridge/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        b"adminSetResourceWithSignature(bytes32,bytes4,uint32,bytes32,address,address,bytes)",
+    )
+    .map_err(|_| ContractError::HashError)?;
+    if msg.function_sig != func_sig[0..4] {
+        return Err(ContractError::Std(StdError::GenericErr {
+            msg: "Invalid function signature".to_string(),
+        }));
+    }
+
+    // Handle the main business
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply contracts/signature-bridge/src/contract.rs]: Done
https://github.com/webb-tools/protocol-cosmwasm/pull/37/commits/2b845e7ee8119bba81d3ff3b693de047f55ba077

**Pull request merged**